File: src/main/java/discretemath/structure/set/multi/SimpleMultiSet.java
package discretemath.structure.set.multi;

import java.util.Collection;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import discretemath.structure.set.Set;
import discretemath.structure.tuple.Pair;
import discretemath.structure.tuple.SimplePair;
import discretemath.utils.CacheValue;

public class SimpleMultiSet<T> extends AbstractMultiSet<T> {

	private final Map<T, Integer> countMap;

	private final int size;

	private final CacheValue<T> modeCache = new CacheValue<>();

	private SimpleMultiSet(Map<T, Integer> countMap) {
		this.countMap = countMap;
		this.size = countMap.values().stream().mapToInt(value -> value).sum();
	}

	public static <T> MultiSet<T> fromCountMap(Map<T, Integer> map) {
		return new SimpleMultiSet<>(Map.copyOf(map));
	}

	public static <T> MultiSet<T> fromCollection(Collection<T> collection) {
		Map<T, Integer> countMap = new HashMap<>();
		for (T item : collection) {
			countMap.merge(item, 1, (oldValue, newValue) -> oldValue + 1);
		}

		return new SimpleMultiSet<>(countMap);
	}

	@Override
	public T mode() {
		synchronized (modeCache) {
			if (modeCache.isSet()) {
				return modeCache.get();
			}

			Iterator<Map.Entry<T, Integer>> iterator = countMap.entrySet().iterator();
			Map.Entry<T, Integer> modeEntry = iterator.next();

			while (iterator.hasNext()) {
				Map.Entry<T, Integer> entry = iterator.next();
				if (entry.getValue() > modeEntry.getValue()) {
					modeEntry = entry;
				}
			}

			T mode = modeEntry.getKey();

			modeCache.set(mode);

			return mode;
		}
	}

	@Override
	public int itemCount(T item) {
		Integer count = countMap.get(item);
		if (count == null) {
			return 0;
		}

		return count;
	}

	@Override
	public int size() {
		return size;
	}

	@Override
	public boolean contains(T item) {
		return countMap.getOrDefault(item, 0) != 0;
	}

	@Override
	public Set<T> unionWith(Set<T> set) {
		Map<T, Integer> resultCountMap = new HashMap<>(countMap);

		if (set instanceof MultiSet<T> multiSet) {
			for (T item : multiSet) {
				int count = multiSet.itemCount(item);
				resultCountMap.merge(item, count, Math::max);
			}
		} else {
			for (T item : set) {
				resultCountMap.merge(item, 1, Math::max);
			}
		}

		return new SimpleMultiSet<>(resultCountMap);
	}

	@Override
	public Set<T> intersectWith(Set<T> set) {
		Map<T, Integer> resultCountMap = new HashMap<>(countMap);

		if (set instanceof MultiSet<T> multiSet) {
			for (Map.Entry<T, Integer> entry : resultCountMap.entrySet()) {
				T item = entry.getKey();
				Integer count = entry.getValue();
				int countInOther = multiSet.itemCount(item);
				entry.setValue(Math.min(count, countInOther));
			}
		} else {
			Iterator<Map.Entry<T, Integer>> iterator = resultCountMap.entrySet().iterator();
			while (iterator.hasNext()) {
				Map.Entry<T, Integer> entry = iterator.next();
				T item = entry.getKey();
				if (set.contains(item)) {
					entry.setValue(1);
				} else {
					iterator.remove();
				}
			}
		}

		return new SimpleMultiSet<>(resultCountMap);
	}

	@Override
	public Set<T> differenceWith(Set<T> set) {
		Map<T, Integer> resultCountMap = new HashMap<>(countMap);

		if (set instanceof MultiSet<T> multiSet) {
			for (T item : multiSet) {
				int count = multiSet.itemCount(item);
				resultCountMap.computeIfPresent(item, (key, value) -> Math.max(0, value - count));
			}
		} else {
			for (T item : set) {
				resultCountMap.computeIfPresent(item, (key, value) -> Math.max(0, value - 1));
			}
		}

		return new SimpleMultiSet<>(resultCountMap);
	}

	@Override
	public MultiSet<T> sumWith(MultiSet<T> set) {
		Map<T, Integer> resultCountMap = new HashMap<>(countMap);

		for (T item : set) {
			int count = set.itemCount(item);
			resultCountMap.merge(item, count, Integer::sum);
		}

		return new SimpleMultiSet<>(resultCountMap);
	}

	@Override
	public Set<T> symmetricDifferenceWith(Set<T> set) {
		throw new UnsupportedOperationException("Not implemented yet"); // TODO
	}

	@Override
	public <P> Set<Pair<T, P>> cartesianProductWith(Set<P> set) {
		throw new UnsupportedOperationException("Not implemented yet"); // TODO
	}

	@Override
	public Iterator<T> iterator() {
		return countMap.keySet().iterator();
	}

	@Override
	public Iterator<Pair<T, Integer>> withCountIterator() {
		Iterator<Map.Entry<T, Integer>> iterator = countMap.entrySet().iterator();
		return new Iterator<>() {
			@Override
			public boolean hasNext() {
				return iterator.hasNext();
			}

			@Override
			public Pair<T, Integer> next() {
				Map.Entry<T, Integer> entry = iterator.next();
				return new SimplePair<>(entry.getKey(), entry.getValue());
			}
		};
	}


	@Override
	public String toString() {
		return countMap.toString();
	}
}

File: src/main/scala/discretemath/structure/set/multi/SimpleMultiSet.scala
package discretemath.structure.set.multi

import discretemath.utils.CacheValue

import scala.collection.immutable

class SimpleMultiSet[T] private(val countMap: Map[T, Int]) extends AbstractMultiSet[T] {

  val cardinality: Int = countMap.values.sum
  private val modeCache = new CacheValue[T]

  override def mode: Option[T] = {
    modeCache.synchronized {
      if (modeCache.isSet) return Option(modeCache.get)

      val mode = countMap.maxBy(_._2)._1

      modeCache.set(mode)

      return Option(mode)
    }

  }

  override def itemCount(item: T): Int = countMap.getOrElse(item, 0)

  override def contains(item: T): Boolean = countMap.getOrElse(item, 0) != 0

  override def unionWith[I <: MultiSet[T], R >: MultiSet[T]](set: I): R = {
    val mergerFunction = set match {
      case multiSet: MultiSet[T] => (k: T) => (k, math.max(this.itemCount(k), multiSet.itemCount(k)))
      case _ => (k: T) =>
        (k, math.max(this.itemCount(k), {
          if (set.contains(k)) 1 else 0
        }))
    }

    val resultCountMap = countMap.keySet.view.concat(set.iterator)
      .map(mergerFunction)
      .toMap

    new SimpleMultiSet[T](resultCountMap)
  }

  override def intersectWith[I <: MultiSet[T], R >: MultiSet[T]](set: I): R = {
    val mergerFunction = set match {
      case multiSet: MultiSet[T] => (k: T) => (k, math.min(this.itemCount(k), multiSet.itemCount(k)))
      case _ => (k: T) =>
        (k, math.min(this.itemCount(k), {
          if (set.contains(k)) 1 else 0
        }))
    }

    val resultCountMap = countMap.keySet.view.concat(set.iterator)
      .map(mergerFunction)
      .toMap

    new SimpleMultiSet[T](resultCountMap)
  }

  override def differenceWith[I <: MultiSet[T], R >: MultiSet[T]](set: I): R = {
    val mergerFunction = set match {
      case multiSet: MultiSet[T] => (k: T) => (k, math.max(0, this.itemCount(k) - multiSet.itemCount(k)))
      case _ => (k: T) =>
        (k, math.max(0, this.itemCount(k) - {
          if (set.contains(k)) 1 else 0
        }))
    }

    val resultCountMap = countMap.keySet.view.concat(set.iterator)
      .map(mergerFunction)
      .toMap

    new SimpleMultiSet[T](resultCountMap)
  }

  override def sumWith(set: MultiSet[T]): MultiSet[T] = {
    val resultCountMap = countMap.keySet.view.concat(set.iterator)
      .map((k: T) => (k, this.itemCount(k) + set.itemCount(k)))
      .toMap

    new SimpleMultiSet[T](resultCountMap)
  }

  override def symmetricDifferenceWith[I <: MultiSet[T], R >: MultiSet[T]](set: I) = throw new UnsupportedOperationException("Not implemented yet") // TODO

  override def cartesianProductWith[P](set: MultiSet[P]) = throw new UnsupportedOperationException("Not implemented yet")

  override def iterator: Iterator[T] = countMap.view.flatMap(tuple => Seq.fill(tuple._2)(tuple._1)).iterator

  override def withCountIterator: Iterator[(T, Int)] = countMap.iterator

  override def toString: String = countMap.toString
}

object SimpleMultiSet {
  def apply[T](map: immutable.Map[T, Int]) = new SimpleMultiSet[T](map)

  def apply[T](collection: Iterable[T]): SimpleMultiSet[T] = {
    val countMap = collection.groupMapReduce(identity)(i => 1)((v1, v2) => v1 + v2)
    new SimpleMultiSet[T](countMap)
  }
}